DOC-6940: Document the MOLT Verify tool (#17354)
cockroachdb/docs: src/current/v23.1/molt-verify.md

---
title: Use the MOLT Verify tool
summary: Learn how to use the MOLT Verify tool to check for post-migration data discrepancies.
toc: true
docs_area: migrate
---

> 

Reviewer: We probably need to add a preview tag here

> The MOLT Verify tool checks for post-migration data discrepancies between a source and a target database.

Reviewer 2: ```suggestion
The MOLT Verify tool checks for data discrepancies between a source and a target database during a [database migration](migration-overview.html).
```

> The tool performs the following verifications to ensure post-migration data integrity:

Reviewer 2: Conceptually, I believe data verification is a prerequisite to a successful (i.e., consistent) migration. So I think we should replace "post-migration" with phrasing like "during a migration", "as part of a migration", etc.

Reviewer 2: @awoods187 @cotedm Please tell me if I have this wrong!

Reviewer: I agree--well spotted

- **Table Verification:** Check that the structure of tables between the source database and the target database are the same.
- **Column Definition Verification:** Check that the column names, data types, constraints, nullability, and other attributes between the source database and the target database are the same. 

> - **Row Value Verification:** Check that the actual data in the tables are the same between the source database and the target database.

Reviewer 3: ```suggestion
- **Row Value Verification:** Check that the actual data in the tables is the same between the source database and the target database.
```
just my preference... i know data can technically be plural

## Supported databases

The following databases are currently supported:

> - PostgreSQL

Reviewer: And CockroachDB!

Author: Oh, right 😄

- MySQL

## Use the MOLT Verify tool

> 1. Get the JDBC connection strings for the source and target databases you want to compare.

Reviewer 3: all the numbers in a markdown list should be "1." (vale caught this also)

> 2. Make sure the SQL user running the MOLT verify tool has access to the necessary tables.

Reviewer 2: ```suggestion
2. Make sure the SQL user running MOLT Verify has access to the necessary tables.
```

Reviewer 3: does access mean read privileges? suggest linking to `grant.html#grant-privileges-on-specific-tables-in-a-database`

Author: I think they mean access to the Postgres/MySQL databases

> 3. Install and build the tool: 

Reviewer 3: suggest consolidating these to something like:
```suggestion
3. Run the following commands to install and build the tool: 
```

    Run the following commands on your local machine:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    git clone https://github.com/cockroachdb/molt.git
    ~~~

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cd molt
    ~~~

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    go build .
    ~~~
    
    To cross-compile, use:
    
    {% include_cached copy-clipboard.html %}
    ~~~ shell
    GOOS=linux GOARCH=amd64 go build -v .
    ~~~ 
    
    Run the following command to verify the installation:
	  
    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./molt verify --help
    ~~~

> 4. Run MOLT Verify: 

Reviewer 3: ```suggestion
4. Run MOLT Verify.
```

>     The `molt verify` command takes two or more JDBC connection strings as arguments. You can append a name for easier readability using `<name>===` in front of the JDBC string. The first argument is considered as the "source of truth". 

Reviewer 3: ```suggestion
    The `molt verify` command takes two or more JDBC connection strings as arguments. You can append a name for easier readability using `<name>===` in front of the JDBC string. The first argument is considered the "source of truth". 
```

>     Examples:

Reviewer 3: ```suggestion
```

>     The following example compares a PostgreSQL database with a CockroachDB database (simplified naming for both instances):

Reviewer 3: the "simplified naming" line is confusing. do you mean that the connection strings contain placeholders?

Author: it's actually unnecessary now that I think of it. Deleting it

    
    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./molt verify \
      'pg_truth===postgres://<username>:<password>@url:<port>/<database>' \
      'crdb_compare===postgresql://<username>:<password>@<host>:<port>/<database>?sslmode=verify-full'
    ~~~

    The following example compares a MySQL database with a CockroachDB database (simplified naming for both instances):

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    ./molt verify \
      'mysql===jdbc:mysql://root@tcp(<host>:<port>)/<database>' \         
      'postgresql://<username>:<password>@<host>:<port>/<database>?sslmode=verify-full'
    ~~~

    Optionally, you can use [supported flags](#supported-flags) to customize the verification results.

5. Review the verification results:

>     Running the MOLT Verify tool will show if there are any missing rows or extraneous tables in the target database. If any data is missing, you can add the missing data to the target database and run `./molt verify` again.

Reviewer 3: is there anything we can link to for "add the missing data"?

Author: ah, good catch. Linking to the INSERT statement doc.

    {{site.data.alerts.callout_info}} 

>     Be aware of data type differences. For example, if your source MySQL table uses an auto-incrementing ID, MOLT will identify a difference in the table definitions when comparing with CockroachDB's UUID type. In such cases, you might have to perform extra steps such as creating composite types within the target database that use the auto-incrementing ID and other types to maintain referential integrity.

Reviewer 3: ```suggestion
    Be aware of data type differences. For example, if your source MySQL table uses an auto-incrementing ID, MOLT will identify a difference in the table definitions when comparing with CockroachDB's UUID type. In such cases, you might have to perform extra steps, such as [creating composite types](create-type.html#create-a-composite-data-type) within the target database that use the auto-incrementing ID and other types to maintain referential integrity.
```

    {{site.data.alerts.end}}

## Supported flags

Flag | Description
----------|------------

> --concurrency int | Number of shards to process at a time. <br>Default value: 16 <br>For faster verification, set this flag to a higher value. <br>Note: Table splitting by shard only works for `int`, `uuid`, and `float` data types.

Reviewer 3: i think the flags should all be in code format

--table_splits int | Number of shards to split the table into. <br>Default value: 16
--row_batch_size int | Number of rows to get from a table at a time. <br>Default value: 20000

## Limitations

- While verifying data, the tool pages 20,000 rows at a time by default, and row values can change in between, which can lead to temporary inconsistencies in data. You can change the row batch size using the `--row_batch_size int` [flag](#supported-flags).
- MySQL enums and set types are not supported.

> - The tool only supports comparing one MySQL database vs a whole CockroachDB schema (which is assumed to be "public").

Reviewer 3: ```suggestion
- The tool only supports comparing one MySQL database to a whole CockroachDB schema (which is assumed to be "public").
```

> - The tool does not handle schema changes between commands well.

Reviewer 2: It may not be clear what this means. @otan can you provide more specifics?

The tool performs the following verifications to ensure post-migration data integrity:

- **Table Verification:** Check that the structure of tables between the source database and the target database are the same.
- **Column Definition Verification:** Check that the column names, data types, constraints, nullability, and other attributes between the source database and the target database are the same. 
- **Row Value Verification:** Check that the actual data in the tables are the same between the source database and the target database.

## Supported databases

The following databases are currently supported:

- PostgreSQL
- MySQL

## Use the MOLT Verify tool

1. Get the JDBC connection strings for the source and target databases you want to compare.
2. Make sure the SQL user running the MOLT verify tool has access to the necessary tables.
3. Install and build the tool: 

    Run the following commands on your local machine:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    git clone https://github.com/cockroachdb/molt.git
    cd molt
    go build .
    ~~~
    
    To cross-compile, use:
    
    {% include_cached copy-clipboard.html %}
    ~~~ shell

>     GOOS=linux GOARCH=amd64 go build -v -o artifacts/molt .

Reviewer 4: nit: remove `-o artifacts/molt` and it'll be in the same directory
